crates/rbuilder/src/building/order_commit.rs

```diff
@@ -754,6 +764,215 @@ impl<'a, 'b, 'c, 'd, Tracer: SimulationTracer> PartialBlockFork<'a, 'b, 'c, 'd,
+                                        );
+                                    self.rollback(rollback_point.clone());
+                                    continue;
+                                }
+
+                                // Now execute the original transaction.
+                                let retry_result = self.commit_tx(
+                                    tx_with_blobs,
+                                    sponsor_ok.cumulative_gas_used,
+                                    gas_reserved,
+                                    sponsor_ok.cumulative_blob_gas_used,
+                                )?;
+
+                                if let Ok(res) = retry_result {
+                                    // Ensure bundle stays profitable after sponsorship.
+                                    let current_coinbase_balance = self.coinbase_balance()?;
+                                    let bundle_profit_so_far = current_coinbase_balance
+                                        .saturating_sub(coinbase_balance_before);
+
+                                    if bundle_profit_so_far >= total_sponsor_amount {
+                                        if !res.tx_info.receipt.success {
+                                            if bundle.dropping_tx_hashes.contains(&tx_hash) {
+                                                // Sponsorship successful but tx reverted (dropping allowed)
+                                                self.rollback(rollback_point.clone());
+                                                continue;
+                                            }
+                                            if !bundle.reverting_tx_hashes.contains(&tx_hash) {
+                                                // Sponsorship successful but tx reverted (not allowed)
+                                                return Ok(Err(BundleErr::TransactionReverted(
+                                                    tx_hash,
+                                                )));
+                                            }
+                                        }
+
+                                        // First the sponsor tx, then the user tx.
+                                        Self::accumulate_tx_execution(sponsor_ok, &mut insert);
+                                        // Extract values before moving res
+                                        let tx_success = res.tx_info.receipt.success;
+                                        let tx_coinbase_profit = res.tx_info.coinbase_profit;
+                                        Self::accumulate_tx_execution(res, &mut insert);
+
+                                        // Update refundable_profit based on ACTUAL transaction profits, not optimistic bundle profit
+                                        // Only add profit from the original transaction if it's refundable and successful
+                                        if bundle.is_tx_refundable(&tx_hash)
```

Comment:
I wonder if all of these can be simplified dramatically. Currently it adds a new branch for committing sponsored txs while it should be possible to do it in the same way so things like coinbase profit and refundable profit are handled by the same code that its handled now. Otherwise its a bit hard to verify that its correct. 

For example, (see comment about `continue` above) I think this approach can lead to part of the bundle being committed while we should revert that bundle instead.

In other words, better approach would be so the sponsorhip path hits `let result = self.commit_tx(`